Q1. What's the difference between a Series and a DataFrame? Give one operation that works on both,
and one that works only on a DataFrame.

     Series is one dimensional (1D) labelled array which stores the data with one column and index.
     Dataframe is two dimensional (2D) things which stores the values in the form of rows and columns.

     There are too many operation that works on both such as sum(),mean(),max()
     sum() it count the sum of all the digits in the series and the datframe.
     mean() it counts the avearge of all the values present in the series and the dataframe.
     max() it counts the maximum values which is present in series and the dataframe.

Q2. What does inplace=True do? Why is the pandas team discouraging its use in newer versions? Give
one example where you'd avoid it.


     inplace=True modifies the existing pandas dataframe or series instead of creating or returning a modified object. 
     pandas team discourging this due to fear loss of data,corrupt of data
     I think it should be avoid when we are working with the some important data also data with that data also if we modifies it there 
     will be no chances to remodify it

Q3. Explain the difference between .fillna() and .dropna(). When would a data analyst pick one over the
other? Be specific - give a concrete situation for each.


     .fillna() is used to fill the values in the columns or rows according to the condition given the round braces where the the data 
     is missing the table.it means that if column has  na (nan) values where this method is used.

     This method is really very helpful when we have less amount of data cause if we do not use this method then we have to remove all the rows 
      and column where this nan value is present..

     .dropna() method is used to drop the column or rows wherever the nan values is present in it.

     it may cause the data loss this is affordable for the data analyst if it is dealing with the small amount of dataset.

     

Q4. What does pd.to_numeric(col, errors='coerce') do that plain pd.to_numeric(col) does NOT? Which
one would you use on real-world messy data, and why?


    pd.to_numeric(col) it tries to converts all the values into the numbers if something comes like which is not able to convert to numbers 
    then it will raise an error and program stops.

    pd.to_numeric(col, errors='coerce') this function handle that all the error which is occuring and converts all the text data into NaN values.

Q5. Explain .loc vs .iloc with one example each. Then describe a case where df.loc[0] and df.iloc[0]
return different rows.

    .loc method used to retrive the values which are present at the labels.

    .iloc is used to retrive the values which are present at the integer position and it does not use the values which is defined by us.

    df.loc[0] returns the different values cause it works with label.

    df.iloc[0] returns the different values cause it works with the integer position.

    





Q6. What is the difference between .map(), .apply(), and .replace() for a Series? When would you reach
for each?

    .map() function is used to assign each value to another values it usually used in dictionary.

    .apply() method used to apply all the changes to all the function which are mentioned in the braces.
    .replace() method used to replace the values with other values.

    .map(),.apply(),.replace() function have different application based on the condition as per your need have to reach out for the each.
    



Q7. What is the SettingWithCopyWarning? Give one example of code that triggers it and one way to fix
it. (Hint: the chained-assignment pattern.)

    i don't know answer for this question....



    

Q8. Explain pd.cut() vs pd.qcut(). If you have annual income data and want to label customers as Low /
Mid / High earners, which one fits better? Why?


    pd.cut() function creates a bins based on fixed value ranges that you specify..


    pd.qcut() function creates a bins based on quantiles so based on number of observations.


    if we have to fit the income range in the category then we have to use pd.cut() function cause in this function we can give a range of income of
    employee and creates a bins and labels like "low","mediam","high" and assign it values correctly.



    

In [78]:
#Q9. Predict the output. What's in df at the end?
import pandas as pd
df = pd.DataFrame({'a': [1, 2, 3], 'b': [4, 5, 6]})
df2 = df
df2['a'] = [10, 20, 30]
print(df)

#output= a:[10,20,30],b:[4,5,6]


    a  b
0  10  4
1  20  5
2  30  6


In [79]:
#Q10. Predict both outputs.
import pandas as pd
import numpy as np
s = pd.Series([1, 2, np.nan, 4, 5])
print(s.mean())
print(s.sum())

# 2.4 wrong
# 12


3.0
12.0


In [80]:
#Q11. Predict the dtype and the sum.
import pandas as pd
df = pd.DataFrame({'price': ['100', '200', 'NA', '400']})
df['price'] = pd.to_numeric(df['price'], errors='coerce')
print(df['price'].dtype)
print(df['price'].sum())

#float
#700

float64
700.0


In [81]:
#Q12. Predict the type of df.loc[0].
import pandas as pd
df = pd.DataFrame({'name': ['Rahul', 'Priya'], 'age': [22, 21]})
print(df.loc[0])
print(type(df.loc[0]))

#Rahul 22
#obj

name    Rahul
age        22
Name: 0, dtype: object
<class 'pandas.core.series.Series'>


In [82]:
#Q13. Predict – and explain what's wrong with the modification line.
import pandas as pd
df = pd.DataFrame({'rating': [4.5, 3.8, 4.2, 5.0]})
df_filtered = df[df['rating'] > 4]
df_filtered['rating'] = df_filtered['rating'] * 2 # what warning?
print(df)
print(df_filtered)

"""
rating
4.5
3.8
4.2
5.0

"""

"""
rating
9
8.4
10

"""


   rating
0     4.5
1     3.8
2     4.2
3     5.0
   rating
0     9.0
2     8.4
3    10.0


C:\Users\omkars\AppData\Local\Temp\ipykernel_12344\108734041.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_filtered['rating'] = df_filtered['rating'] * 2 # what warning?


'\nrating\n9\n8.4\n10\n\n'

In [83]:
#Q14. Predict the result AND its dtype. Why is the dtype what it is?
import pandas as pd
df = pd.DataFrame({'a': [1, 2, 3, 4]})
result = df['a'].apply(lambda x: x**2 if x > 2 else 'small')
print(result)
print(result.dtype)

# [small,small,9,16]
#obj

0    small
1    small
2        9
3       16
Name: a, dtype: object
object


In [84]:
#Q15. Predict each output. What's the difference between nunique(), len(unique()), and value_counts()?
import pandas as pd
df = pd.DataFrame({'city': ['Delhi', 'Mumbai', 'Delhi', 'Bengaluru', None]})
print(df['city'].nunique())
print(len(df['city'].unique()))
print(df['city'].value_counts())

#3
#4
#Delhi 2
#Mumbai 1
#Bengaluru 1

3
4
city
Delhi        2
Mumbai       1
Bengaluru    1
Name: count, dtype: int64


In [85]:
import pandas as pd
df = pd.read_csv("indian_startups_funding.csv")
# df.head()
# df["City"].value_counts()

df = df.rename(columns={"Startup Name ": "startup_name","City":"city","Funding Stage": "funding_stage","INDUSTRY":"industry","Funding Date":"funding_date","Founded Year":"founded_year","Investors":"investors","Amount":"amount"})

#df.rename(columns={"Startup Name":"startup_name"})
df["city"] = df["city"].str.capitalize().str.strip()

In [86]:
df["city"]

0          Delhi
1      Ahmedabad
2        Kolkata
3         Mumbai
4        Kolkata
         ...    
463      Kolkata
464      Kolkata
465      Chennai
466      Kolkata
467     Calcutta
Name: city, Length: 468, dtype: object

In [87]:
df["startup_name"].str.capitalize().str.strip()

0           Playtime games
1              Rebel foods
2               Freshworks
3                Bigbasket
4                      Ola
              ...         
463               Doubtnut
464               Ecomotor
465                  Zepto
466    Fasttrack logistics
467             Freshworks
Name: startup_name, Length: 468, dtype: object

In [88]:
city_mapping = {
    "Bangalore": "Bengaluru",
    "BLR": "Bengaluru",
    "Blr": "Bengaluru",
    
    "Bombay": "Mumbai",
    "Bom": "Mumbai",
    
    "Gurgaon": "Gurugram",
    "Ggn": "Gurugram",
    
    "Delhi": "New Delhi",
    "New delhi": "New Delhi",
    
    "Calcutta": "Kolkata",
    
    "Madras": "Chennai",
    
    "Hyd": "Hyderabad"
}

df["city"] = df["city"].map(city_mapping).fillna(df["city"])

In [89]:
industry_mapping = {
    "AI/ML": "AI/ML",
    "AI-ML": "AI/ML",
    "AI ML": "AI/ML",
    "AIML": "AI/ML",

    "TRAVELTECH": "TravelTech",
    "Travel-Tech": "TravelTech",
    "TravelTech": "TravelTech",

    "BEAUTYTECH": "BeautyTech",
    "BeautyTech": "BeautyTech",
    "Beauty-Tech": "BeautyTech",

    "logistics": "Logistics",
    "LOGISTICS": "Logistics",
    "Logistics": "Logistics",

    "saas": "SaaS",
    "Saas": "SaaS",
    "SAAS": "SaaS",
    "SaaS": "SaaS",

    "PROPTECH": "PropTech",
    "Prop-Tech": "PropTech",
    "PropTech": "PropTech",

    "GAMING": "Gaming",
    "Gaming": "Gaming",
    "gaming": "Gaming",
    "Games": "Gaming",

    "Agri Tech": "AgriTech",
    "AgriTech": "AgriTech",
    "Agri-Tech": "AgriTech",
    "AGRITECH": "AgriTech",
    "agritech": "AgriTech",

    "Health-Tech": "HealthTech",
    "HealthTech": "HealthTech",
    "healthtech": "HealthTech",
    "HEALTHTECH": "HealthTech",
    "Health Tech": "HealthTech",

    "Electric Vehicles": "EV",
    "Electric-Vehicles": "EV",
    "EV": "EV",
    "ev": "EV",

    "FoodTech": "FoodTech",
    "Food-Tech": "FoodTech",
    "Food Tech": "FoodTech",
    "foodtech": "FoodTech",
    "FOOD-TECH": "FoodTech",

    "Edtech": "EdTech",
    "EdTech": "EdTech",
    "EDTECH": "EdTech",
    "Ed-Tech": "EdTech",

    "ecommerce": "E-Commerce",
    "E-Commerce": "E-Commerce",
    "Ecommerce": "E-Commerce",
    "E-commerce": "E-Commerce",
    "ECOMMERCE": "E-Commerce",

    "FINTECH": "FinTech",
    "Fin-Tech": "FinTech",
    "fintech": "FinTech",
    "Fintech": "FinTech",
    "FinTech": "FinTech",
    "FT": "FinTech"
}

df["industry"] = df["industry"].map(industry_mapping).fillna(df["industry"])

In [90]:
funding_stage_mapping = {
    "Seed": "Seed",
    "SEED": "Seed",
    "seed": "Seed",

    "Pre-Seed": "Pre-Seed",
    "Pre Seed": "Pre-Seed",
    "PRE-SEED": "Pre-Seed",
    "Preseed": "Pre-Seed",

    "Pre-Series A": "Pre-Series A",
    "PRE-SERIES A": "Pre-Series A",
    "Pre Series A": "Pre-Series A",

    "Pre-A": "Pre-Series A",

    "Series-A": "Series A",
    "SERIES A": "Series A",
    "Series A": "Series A",
    "series a": "Series A",
    "series_a": "Series A",

    "Series-B": "Series B",
    "SERIES B": "Series B",
    "Series B": "Series B",
    "series_b": "Series B",
    "series b": "Series B",

    "Series-C": "Series C",
    "SERIES C": "Series C",
    "Series C": "Series C",
    "series c": "Series C",

    "Series-D": "Series D",
    "SERIES D": "Series D",
    "Series D": "Series D",
    "series d": "Series D",

    "Series-E": "Series E",
    "SERIES E": "Series E",
    "Series E": "Series E",
    "series e": "Series E",

    "Bridge": "Bridge",
    "bridge": "Bridge",
    "BRIDGE": "Bridge",

    "IPO": "IPO",
    "ipo": "IPO",

    "Pre-IPO": "Pre-IPO"
}


df["funding_stage"] = df["funding_stage"].map(funding_stage_mapping).fillna(df["funding_stage"])

In [91]:
df["amount"] = pd.to_numeric(df["amount"],errors="coerce")

In [92]:
df["amount"].value_counts()

amount
2196.0    2
900.0     2
726.0     2
719.0     2
65.0      2
         ..
15.0      1
35.0      1
493.0     1
1430.0    1
157.0     1
Name: count, Length: 72, dtype: int64

In [93]:
df["funding_date"] = pd.to_datetime(df["funding_date"],errors="coerce",dayfirst=True)

Q17. After loading the raw CSV (before cleaning), answer:
(a) How many rows and columns?
(b) How many missing values are in each column?
(c) How many exact duplicate rows are present?



(a)- 468 rows and 8 columns
(b)- 
startup_name       0
city               0
industry           0
funding_stage      0
amount           389
funding_date     375
investors         10
founded_year      23
dtype: int64

(c)- 18 duplicate rows are present (np.int64(18))

Q18. After cleaning column names to snake_case, print the cleaned column names sorted
alphabetically. Don't manually type them – use Python on df.columns.

    Index(['startup_name', 'city', 'industry', 'funding_stage', 'amount',
       'funding_date', 'investors', 'founded_year'],
      dtype='object')

In [94]:
#Q19. Without using .describe(), compute these statistics on the cleaned founded_year column (after
#dropping NaN): count, mean, min, max, median, std. Print as a single dictionary with these exact keys.

dict1 = {"count":df["founded_year"].count(),
"mean":df["founded_year"].mean(),
"min":df["founded_year"].min(),
"max":df["founded_year"].max(),
"median":df["founded_year"].median(),
"std":df["founded_year"].std()}

dict1

{'count': np.int64(445),
 'mean': np.float64(2016.4606741573034),
 'min': np.float64(2008.0),
 'max': np.float64(2022.0),
 'median': np.float64(2017.0),
 'std': np.float64(4.187218270138972)}

In [95]:
#Q20. After cleaning the city column with the aliases mapped, print all unique city names sorted
#alphabetically. There should be exactly 10 cities and no duplicates from case or whitespace.

df["city"].unique()



array(['New Delhi', 'Ahmedabad', 'Kolkata', 'Mumbai', 'Hyderabad',
       'Noida', 'Pune', 'Bengaluru', 'Chennai', 'Gurugram'], dtype=object)

In [96]:
#Q21. After standardising the industry column using a dict, print the count of startups per industry sorted
#from most to least common (use value_counts()).


df["industry"].value_counts()

industry
AI/ML         56
FoodTech      39
SaaS          38
Gaming        37
BeautyTech    36
PropTech      34
TravelTech    34
HealthTech    33
E-Commerce    33
AgriTech      30
Logistics     27
EV            27
EdTech        22
FinTech       22
Name: count, dtype: int64

In [97]:
"""Q22. After converting amount to numeric, print:
(a) The dtype of the column
(b) How many NaN values are there (from 'Undisclosed' entries)
(c) The mean funding amount (ignoring NaN)
(d) The minimum and maximum funding amount"""


print(df["amount"].dtypes)
print(df["amount"].mean())
print(df["amount"].min())
print(df["amount"].max())

float64
704.5569620253165
13.0
3545.0


In [98]:
"""Q23. After converting funding_date to datetime:
(a) Count how many deals happened in each year. Use .dt.year.
(b) Print the year with the highest number of funding deals.
"""

print(df["funding_date"].dt.year.value_counts())
print(df["funding_date"].dt.year.value_counts().idxmax())



funding_date
2022.0    24
2023.0    14
2021.0    14
2024.0     7
2019.0     7
2020.0     6
2017.0     5
2018.0     5
2016.0     4
2015.0     3
2011.0     1
2014.0     1
2013.0     1
2012.0     1
Name: count, dtype: int64
2022.0


In [99]:
"""Q24. After standardising funding_stage, print the count of startups per stage (sorted from most
common to least)."""

df["funding_stage"].value_counts().sort_index()


funding_stage
Bridge          42
IPO             34
Pre-IPO         12
Pre-Seed        48
Pre-Series A    41
Seed            49
Series A        49
Series B        58
Series C        47
Series D        40
Series E        48
Name: count, dtype: int64

In [100]:
#Q25. Find all Bengaluru-based startups with funding amount > 100 Cr. Print the startup_name and
#amount columns, sorted by amount descending.

In [101]:
df1 = df[(df["city"]=="Bengaluru") & (df["amount"] > 100)]
df1[["startup_name", "amount"]].sort_values("amount",ascending=False)

,startup_name,amount
226,AgriConnect,3526.0
146,Curefoods,2230.0
252,PlayTime Games,1410.0
369,Cred,950.0
149,MediConnect,304.0
398,Cuemath,115.0


In [102]:
#Q26. Using .isin(), find startups whose industry is one of: Fintech, Edtech, HealthTech. Print the total count

count = df[df["industry"].isin(["Fintech", "Edtech", "HealthTech"])]
print(len(count))

33


In [103]:
#Q27. Using .between(), find startups founded between 2015 and 2020 (both inclusive). Print the count.
#Be careful – founded_year has missing values.

year = df[df["founded_year"].between(2015,2020)]
print(year)

            startup_name       city    industry funding_stage  amount  \
0         PlayTime Games  New Delhi  E-Commerce      Series E     NaN   
1            Rebel Foods  Ahmedabad  BeautyTech  Pre-Series A  1440.0   
2             Freshworks    Kolkata    AgriTech  Pre-Series A     NaN   
3              BigBasket     Mumbai  E-Commerce      Series D     NaN   
4                    Ola    Kolkata      EdTech      Series E     NaN   
..                   ...        ...         ...           ...     ...   
461              Postman  Hyderabad      EdTech      Series B     NaN   
462          Rebel Foods  Hyderabad    AgriTech      Series B   157.0   
465                Zepto    Chennai       AI/ML        Bridge     NaN   
466  FastTrack Logistics    Kolkata    AgriTech        Bridge     NaN   
467           Freshworks    Kolkata     FinTech          Seed     NaN   

    funding_date                                          investors  \
0     2024-02-29                        Ranjan Pai, 

In [104]:
#Q28. Using .str.contains() with case=False and na=False, find all startups whose name contains the word
#'pay'. Print the unique names.

uniqure = df[df["startup_name"].str.contains("pay")]
print(len(uniqure))

3


In [106]:
"""Q29. Add a new column funding_efficiency = amount / (2024 - founded_year). Higher = raised more
capital per year of existence. Sort by funding_efficiency descending. Print top 10 with name, industry,
amount, founded_year, and funding_efficiency."""


df["funding_efficiency"] = df["amount"]/ (2024 - df["founded_year"])
sort = df.sort_values("funding_efficiency",ascending=False)

print(sort[[
    "startup_name",
    "industry",
    "amount",
    "founded_year",
    "funding_efficiency"
]].head(10))




    startup_name    industry  amount  founded_year  funding_efficiency
55       Blinkit          EV  2196.0        2022.0         1098.000000
144    GroomGuru      Gaming  3497.0        2019.0          699.400000
146    Curefoods    FoodTech  2230.0        2020.0          557.500000
456     DigiKart    PropTech  1430.0        2021.0          476.666667
56   WhiteHat Jr        SaaS  2250.0        2019.0          450.000000
450       Swiggy  BeautyTech   726.0        2022.0          363.000000
454    CodeCraft  BeautyTech   719.0        2022.0          359.500000
110       Studyo    FoodTech  3545.0        2014.0          354.500000
226  AgriConnect  TravelTech  3526.0        2014.0          352.600000
417  VillageMart      EdTech   896.0        2021.0          298.666667


In [107]:
"""Q30. Using .apply() with a lambda, add a funding_tier column based on amount:
amount ≥ 1000 Cr → 'Unicorn'
amount ≥ 500 Cr → 'Mega'
amount ≥ 100 Cr → 'Large'
amount ≥ 10 Cr → 'Mid'
else (including NaN) → 'Small / Unknown'
Print the count of startups in each tier."""


df["funding_tier"] = df["amount"].apply(
    lambda amount: "Unicorn" if amount >= 1000
    else "Mega" if amount >= 500
    else "Large" if amount >= 100
    else "Mid" if amount >= 10
    else "Small / Unknown"
)

print(df["funding_tier"].value_counts())

funding_tier
Small / Unknown    389
Mid                 26
Large               21
Unicorn             17
Mega                15
Name: count, dtype: int64


In [108]:
"""Q31. Using pd.cut(), bin founded_year into 3 eras:
Before 2010 → 'Early'
2010 to 2017 → 'Growth'
2018 and later → 'Recent'
Print the count of startups in each era."""

result= pd.cut(df["founded_year"],bins=[1900,2009,2017,2024],labels = ["Early","Growth","Recent"])

print(result.value_counts())
      



founded_year
Recent    214
Growth    206
Early      25
Name: count, dtype: int64


In [109]:
"""32. The investors column has multiple investors separated by ', ' (comma + space). Create a new
column investor_count showing the number of investors per startup. Use .apply() with a lambda. Handle
NaN values (count = 0). Print the value_counts() of investor_count."""


df["investor_count"] = df["investors"].apply(
    lambda x: len(x.split(", ")) if pd.notna(x) else 0
)

print(df["investor_count"].value_counts())

investor_count
2    208
3    122
1     98
4     30
0     10
Name: count, dtype: int64


In [110]:
"""Q33. Compute the average funding amount per industry (ignoring NaN). Sort descending. Print the top 5
industries by average funding."""

avg = df.groupby("industry")["amount"].mean().sort_values(ascending=False)

print(avg.head(5))

industry
E-Commerce    1411.000000
TravelTech    1193.500000
HealthTech    1066.000000
SaaS           911.166667
Gaming         848.833333
Name: amount, dtype: float64


In [111]:
"""Q34. For each city, compute three metrics in a single groupby using .agg() with a dictionary:
(a) Number of startups
(b) Total funding raised (sum of amount)
(c) Average funding per startup
Print the top 10 cities by number of startups."""


metrics = {
    "startup_name": "count",
    "amount": ["sum", "mean"]
}

output = df.groupby("city").agg(metrics)

output.columns = ["number_of_startups", "total_funding", "average_funding"]

print(output.sort_values("number_of_startups", ascending=False).head(10))

           number_of_startups  total_funding  average_funding
city                                                         
Noida                      69         5556.0       617.333333
Bengaluru                  51         8724.0       727.000000
Kolkata                    51         4316.0       479.555556
Mumbai                     46          869.0       173.800000
Hyderabad                  46         4578.0       915.600000
Ahmedabad                  45         5942.0       742.750000
Pune                       44         4917.0       546.333333
Gurugram                   44        11850.0      1077.272727
Chennai                    39         4319.0       617.000000
New Delhi                  33         4589.0      1147.250000


In [112]:
"""Q35. Interview-style. Which industry has the highest average funding, but only considering industries
with at least 15 startups? (Sparse industries bias the mean.) Print industry name, count, and average
funding."""

output = df.groupby("industry").agg(count = ("startup_name","count"),
                                   average_funding = ("amount","mean"))
output = output[output["count"] >= 15]

output = output.sort_values("count",ascending=True)

print(output)

            count  average_funding
industry                          
FinTech        22       306.000000
EdTech         22       671.333333
EV             27       828.888889
Logistics      27       625.166667
AgriTech       30       103.200000
E-Commerce     33      1411.000000
HealthTech     33      1066.000000
PropTech       34       411.000000
TravelTech     34      1193.500000
BeautyTech     36       688.666667
Gaming         37       848.833333
SaaS           38       911.166667
FoodTech       39       832.900000
AI/ML          56       372.888889


In [113]:
"""Q36. Using pd.crosstab(), show the distribution of funding_tier across the top 5 industries by count.
Rows = industry, columns = funding_tier, values = count of startups."""

output = pd.crosstab(df["industry"],df["funding_tier"])
E = df["industry"].value_counts().head(5).index

output = output.loc[E]
print(output)

funding_tier  Large  Mega  Mid  Small / Unknown  Unicorn
industry                                                
AI/ML             2     1    5               47        1
FoodTech          4     2    2               29        2
SaaS              1     1    2               32        2
Gaming            0     0    4               31        2
BeautyTech        4     2    1               27        2


In [114]:
"""Q37. For each industry, find the startup with the highest funding amount. Print startup_name, industry,
city, and amount. Hint: groupby + idxmax() on amount."""


index = df.groupby("industry")["amount"].idxmax()

High_funding_amount = df.loc[index,["startup_name", "industry", "city", "amount"]]

print(High_funding_amount)

      startup_name    industry       city  amount
134      Games24x7       AI/ML  Ahmedabad  1406.0
257       EduSmart    AgriTech      Noida   193.0
284       DigiKart  BeautyTech    Chennai  2249.0
312  BookMyService  E-Commerce  Hyderabad  1411.0
12             Ola          EV  New Delhi  3520.0
369           Cred      EdTech  Bengaluru   950.0
231      FreshDeli     FinTech   Gurugram   306.0
110         Studyo    FoodTech   Gurugram  3545.0
144      GroomGuru      Gaming      Noida  3497.0
373     TripScribe  HealthTech       Pune  1421.0
268       DataLake   Logistics  Ahmedabad  2196.0
456       DigiKart    PropTech   Gurugram  1430.0
56     WhiteHat Jr        SaaS  Hyderabad  2250.0
226    AgriConnect  TravelTech  Bengaluru  3526.0


In [115]:
"""Q38. Emerging player finder. An 'emerging player' is a startup founded in 2019 or later, with funding ≥
100 Cr, in one of Fintech / Edtech / HealthTech. Find all of them, sort by amount descending, print
name, industry, city, founded_year, and amount. Why might a VC analyst care about this list?"""



emerging_players = df[(df["founded_year"] >= 2019) & (df["amount"] >= 100) & (df["industry"].isin(["Fintech", "Edtech", "HealthTech"]))]

emerging_players = emerging_players.sort_values("amount",ascending=False)

print(emerging_players[["startup_name", "industry", "city", "founded_year", "amount"]])

    startup_name    industry      city  founded_year  amount
130    DermaPlus  HealthTech  Gurugram        2019.0   711.0


In [116]:
"""Q39. Investor hub analysis. A foreign investor wants to know which Indian city has the highest TOTAL
funding raised, considering only cities with at least 20 startups (so the number is meaningful). Print city
name, count, and total funding."""

output = df.groupby("city").agg(
    count=("startup_name", "count"),
    total_funding=("amount", "sum")
)

output = output[output["count"] >= 20]

output = output.sort_values("total_funding", ascending=False)

print(output)

           count  total_funding
city                           
Gurugram      44        11850.0
Bengaluru     51         8724.0
Ahmedabad     45         5942.0
Noida         69         5556.0
Pune          44         4917.0
New Delhi     33         4589.0
Hyderabad     46         4578.0
Chennai       39         4319.0
Kolkata       51         4316.0
Mumbai        46          869.0


In [117]:
"""Q40. Market saturation. For each combination of (city, industry), count the number of startups. Print the
top 10 most-saturated combinations. This tells a new founder where competition is most intense – and
by elimination, where the gaps might be."""


outpu = df.groupby(["city","industry"]).size().sort_values(ascending=True)

print(outpu)

city       industry  
Ahmedabad  FinTech        1
           EV             1
Chennai    EV             1
           EdTech         1
Gurugram   Logistics      1
                         ..
Mumbai     AI/ML          8
Pune       AI/ML          8
Noida      BeautyTech     8
Bengaluru  Gaming         9
Gurugram   FoodTech      10
Length: 134, dtype: int64
